agent/agent.go

```diff
@@ -0,0 +1,825 @@
+}
+
+func (o *opAgent) state() (bool, string, string) {
+	o.RLock()
+	defer o.RUnlock()
+	return o.Running(), o.executablePath, o.configPath
+}
+
+func (o *opAgent) Start(ctx context.Context, request *m3em.StartRequest) (*m3em.StartResponse, error) {
+	o.logger.Infof("received Start()")
+	o.Lock()
+	defer o.Unlock()
+
+	if o.Running() {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "already running")
+	}
+
+	if o.executablePath == "" {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "agent missing build")
+	}
+
+	if o.configPath == "" {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "agent missing config")
+	}
+
+	if err := o.startWithLock(); err != nil {
+		return nil, grpc.Errorf(codes.Internal, "unable to start: %v", err)
+	}
+
+	return &m3em.StartResponse{}, nil
+}
+
+func (o *opAgent) newListenerFn() func(error) {
+	return func(err error) {
+		if err == nil {
+			err = fmt.Errorf("test process terminated without error")
+		} else {
+			err = fmt.Errorf("test process terminated with error: %v", err)
+		}
+		o.logger.Warnf("%v", err)
+		if stopping := atomic.LoadInt32(&o.stopping); stopping == 0 && o.heartbeater != nil {
+			o.heartbeater.notifyProcessTermination(err)
+		}
+		atomic.StoreInt32(&o.running, 0)
+	}
+}
+
+func (o *opAgent) newProcessListener() exec.ProcessListener {
+	var (
+		listenFn   = o.newListenerFn()
+		onComplete = func() { listenFn(nil) }
+		onError    = func(err error) { listenFn(err) }
+	)
+	return exec.NewProcessListener(onComplete, onError)
+}
+
+func (o *opAgent) startWithLock() error {
+	var (
+		path, args = o.opts.ExecGenFn()(o.executablePath, o.configPath)
+		osArgs     = append([]string{path}, args...)
+		cmd        = exec.Cmd{
+			Path:      path,
+			Args:      osArgs,
+			OutputDir: o.opts.WorkingDirectory(),
+			Env:       o.opts.EnvMap(),
+		}
+		listener = o.newProcessListener()
+	)
+	pm, err := o.newProcessMonitorFn(cmd, listener)
+	if err != nil {
+		return err
+	}
+	o.logger.Infof("executing command: %+v", cmd)
+	if err := pm.Start(); err != nil {
+		return err
+	}
+	atomic.StoreInt32(&o.running, 1)
+	o.processMonitor = pm
+	return nil
+}
+
+func (o *opAgent) Stop(ctx context.Context, request *m3em.StopRequest) (*m3em.StopResponse, error) {
+	o.logger.Infof("received Stop()")
+	o.Lock()
+	defer o.Unlock()
+
+	if !o.Running() {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "not running")
+	}
+
+	atomic.StoreInt32(&o.stopping, 1)
+	if err := o.stopWithLock(); err != nil {
+		return nil, grpc.Errorf(codes.Internal, "unable to stop: %v", err)
+	}
+	atomic.StoreInt32(&o.stopping, 0)
+
+	return &m3em.StopResponse{}, nil
+}
+
+func (o *opAgent) stopWithLock() error {
+	if o.processMonitor == nil {
+		return errProcessMonitorNotDefined
+	}
+
+	if err := o.processMonitor.Stop(); err != nil {
+		return err
+	}
+
+	o.processMonitor = nil
+	atomic.StoreInt32(&o.running, 0)
+	return nil
+}
+
+func (o *opAgent) resetWithLock(heartbeatNotification error) error {
```

Comment:
For sure. 

`resetWithLock` is triggered when the agent receives a Setup, a Teardown, or if the agent hasn't been able to heartbeat past a defined timeout. 

In the first two cases, we might be heartbeating to a remote process already. So we use `heartbeatNotification` to send a message back to the process to let them know they've been overwritten. 

We don't do so in the third case (heartbeating has timed out) as we wouldn't be able to send the message back in that case. 

Does using an error for this seem un-idiomatic?